pythran/pythonic/numpy/fft/rfft.hpp

```diff
@@ -1,145 +1,93 @@
+    types::ndarray<typename std::enable_if<std::is_floating_point<T>::value, std::complex<T>>::type,
+            types::array<long, std::tuple_size<pS>::value>>
+    rfft(types::ndarray<T, pS> const &in_array, types::none_type n, long axis,
+          types::none_type norm)
+    {
+        return r2c(in_array, -1, axis, "", true, false);
+    }
 
-      double *rptr = (double *)out_array.buffer;
-      long nrepeats = in_array.flat_size() / npts;
-      long to_copy = (NFFT <= npts) ? NFFT : npts;
-      for (i = 0; i < nrepeats; i++) {
-        rptr[2 * out_size - 1] = 0.0; // We didn't zero the array upon
-                                      // allocation. Make sure the last element
-                                      // is 0.
-        std::copy(dptr, dptr + to_copy, rptr + 1);
-        // Zero padding if the FFT size is > the number points
-        std::fill(rptr + 1 + to_copy, rptr + 1 + to_copy + (NFFT - to_copy), 0);
-        npy_rfftf(NFFT, rptr + 1, twiddle_buffer);
-        rptr[0] = rptr[1];
-        rptr[1] = 0.0;
-        rptr += 2 * out_size;
-        dptr += npts;
-      }
-      if (norm) {
-        double scale = 1. / sqrt(NFFT);
-        rptr = (double *)out_array.buffer;
-        long count = 2 * out_array.flat_size();
-        // Remember that these are complex numbers!
-        for (i = 0; i < count; i++) {
-          rptr[i] *= scale;
-        }
-      }
-      return out_array;
+    template <class T , class pS>
+    types::ndarray<typename std::enable_if<std::is_floating_point<T>::value, std::complex<T>>::type,
+            types::array<long, std::tuple_size<pS>::value>>
+    rfft(types::ndarray<T, pS> const &in_array, long n, long axis,
+          types::none_type norm)
+    {
+        return r2c(in_array, n, axis, "", true, false);
     }
 
-    template <class T, class pS>
-    types::ndarray<std::complex<typename std::common_type<T, double>::type>,
-                   types::array<long, std::tuple_size<pS>::value>>
-    rfft(types::ndarray<T, pS> const &in_array, long NFFT, long axis)
+    template <class T , class pS>
+    types::ndarray<typename std::enable_if<std::is_floating_point<T>::value, std::complex<T>>::type,
+            types::array<long, std::tuple_size<pS>::value>>
+    rfft(types::ndarray<T, pS> const &in_array, long n, long axis,
+          types::str const &norm)
     {
-      return rfft(in_array, NFFT, axis, "");
+        return r2c(in_array, n, axis, norm, true, false);
     }
 
-    // This is kludgy, and I'm sure there's a better way to do this. Jeanl
-    bool testThis(types::none_type param)
+    template <class T , class pS>
+    types::ndarray<typename std::enable_if<std::is_integral<T>::value, std::complex<double>>::type,
+            types::array<long, std::tuple_size<pS>::value>>
+    rfft(types::ndarray<T, pS> const &in_array, types::none_type n, long axis,
+          types::str const &norm)
     {
-      return false;
+        auto tmp_array = _copy_to_double(in_array);
```

Comment:
See my comment above I think the code is more explicit (albeit a bit more code) if 
we overload for integer and floating instead of merging them and using conditionals.